### PR TITLE
feat: ae2 empty interface compatibility (item + fluids) and some other refactoring

### PR DESCRIPTION
**Pull request by Vonr**

### CODE REVIEW

File: src/main/java/ca/teamdman/sfm/common/capabilityprovidermapper/ae2/InterfaceCapabilityProviderMapper.java [hunk 1]
```diff
@@ -0,0 +1,446 @@
...
+            if (in != null) {
+                return in.getCapability(cap, side);
+            }
+
+            return LazyOptional.empty();
+        }
+    }
+
+    @Nullable
+    private static InterfaceBlockEntity interfaceAt(LevelAccessor level, BlockPos pos) {
```
Comment (TeamDman) [inline src/main/java/ca/teamdman/sfm/common/capabilityprovidermapper/ae2/InterfaceCapabilityProviderMapper.java]:
```suggestion
    private static @Nullable InterfaceBlockEntity interfaceAt(LevelAccessor level, BlockPos pos) {
```

File: src/main/java/ca/teamdman/sfm/common/capabilityprovidermapper/ae2/InterfaceCapabilityProviderMapper.java [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+        var be = level.getBlockEntity(pos);
+        if (be instanceof InterfaceBlockEntity in) {
+            return in;
+        }
+        return null;
+    }
+
+    record InterfaceHandler(LevelAccessor level, BlockPos pos) implements IItemHandler, IFluidHandler {
+        @Nullable
+        InterfaceBlockEntity getInterface() {
```
Comment (TeamDman) [inline src/main/java/ca/teamdman/sfm/common/capabilityprovidermapper/ae2/InterfaceCapabilityProviderMapper.java]:
```suggestion
        @Nullable InterfaceBlockEntity getInterface() {
```

File: src/main/java/ca/teamdman/sfm/common/capabilityprovidermapper/ae2/InterfaceCapabilityProviderMapper.java [hunk 3]
```diff
@@ -0,0 +1,446 @@
...
+            return interfaceAt(this.level, this.pos);
+        }
+
+        @Nullable
+        IEnergyService getEnergy() {
```
Comment (TeamDman) [inline src/main/java/ca/teamdman/sfm/common/capabilityprovidermapper/ae2/InterfaceCapabilityProviderMapper.java]:
```suggestion
        @Nullable IEnergyService getEnergy() {
```

File: src/main/java/ca/teamdman/sfm/common/capabilityprovidermapper/ae2/InterfaceCapabilityProviderMapper.java [hunk 4]
```diff
@@ -0,0 +1,446 @@
...
+            var grid = in.getMainNode().getGrid();
+            if (grid == null) {
+                return null;
+            }
+
+            return grid.getEnergyService();
+        }
+
+        @NotNull
+        LazyOptional<IStorageMonitorableAccessor> getCapability() {
```
Comment (TeamDman) [inline src/main/java/ca/teamdman/sfm/common/capabilityprovidermapper/ae2/InterfaceCapabilityProviderMapper.java]:
```suggestion
        LazyOptional<IStorageMonitorableAccessor> getCapability() {
```
the package file should be adding notnullbydefault to everything

File: src/main/java/ca/teamdman/sfm/common/capabilityprovidermapper/ae2/InterfaceCapabilityProviderMapper.java [hunk 5]
```diff
@@ -0,0 +1,446 @@
...
+
+            if (slots == null) {
+                slots = this.withBaseItemHandler(IItemHandler::getSlots);
+            }
+
+            return slots == null ? 0 : slots;
+        }
+
+        @Override
+        public @NotNull ItemStack getStackInSlot(int slot) {
```
Comment (TeamDman) [inline src/main/java/ca/teamdman/sfm/common/capabilityprovidermapper/ae2/InterfaceCapabilityProviderMapper.java]:
```suggestion
        public ItemStack getStackInSlot(int slot) {
```
the package file should be adding notnullbydefault to everything

File: src/main/java/ca/teamdman/sfm/common/capabilityprovidermapper/ae2/InterfaceCapabilityProviderMapper.java [hunk 6]
```diff
@@ -0,0 +1,446 @@
...
+
+            if (stack == null) {
+                stack = this.withBaseItemHandler(c -> c.getStackInSlot(slot));
+            }
+
+            return stack == null ? ItemStack.EMPTY : stack;
+        }
+
+        @Override
+        public @NotNull ItemStack insertItem(int slot, @NotNull ItemStack stack, boolean simulate) {
```
Comment (TeamDman) [inline src/main/java/ca/teamdman/sfm/common/capabilityprovidermapper/ae2/InterfaceCapabilityProviderMapper.java]:
```suggestion
        public ItemStack insertItem(int slot, @NotNull ItemStack stack, boolean simulate) {
```
the package file should be adding notnullbydefault to everything

File: src/main/java/ca/teamdman/sfm/common/capabilityprovidermapper/ae2/InterfaceCapabilityProviderMapper.java [hunk 7]
```diff
@@ -0,0 +1,446 @@
...
+                return stack;
+            }
+
+            var rtn = stack.copy();
+            rtn.shrink(inserted);
+            return rtn;
+        }
+
+        @Override
+        public @NotNull ItemStack extractItem(int slot, int amount, boolean simulate) {
```
Comment (TeamDman) [inline src/main/java/ca/teamdman/sfm/common/capabilityprovidermapper/ae2/InterfaceCapabilityProviderMapper.java]:
```suggestion
        public ItemStack extractItem(int slot, int amount, boolean simulate) {
```
the package file should be adding notnullbydefault to everything

File: src/main/java/ca/teamdman/sfm/common/capabilityprovidermapper/ae2/InterfaceCapabilityProviderMapper.java [hunk 8]
```diff
@@ -0,0 +1,446 @@
...
+
+            if (slots == null) {
+                slots = this.withBaseFluidHandler(IFluidHandler::getTanks);
+            }
+
+            return slots == null ? 0 : slots;
+        }
+
+        @Override
+        public @NotNull FluidStack getFluidInTank(int tank) {
```
Comment (TeamDman) [inline src/main/java/ca/teamdman/sfm/common/capabilityprovidermapper/ae2/InterfaceCapabilityProviderMapper.java]:
```suggestion
        public FluidStack getFluidInTank(int tank) {
```

File: src/main/java/ca/teamdman/sfm/common/capabilityprovidermapper/ae2/InterfaceCapabilityProviderMapper.java [hunk 9]
```diff
@@ -0,0 +1,446 @@
...
+
+            if (inserted == null) {
+                inserted = this.withBaseFluidHandler(c -> c.fill(resource, action));
+            }
+
+            return inserted == null ? 0 : inserted;
+        }
+
+        @Override
+        public @NotNull FluidStack drain(FluidStack resource, FluidAction action) {
```
Comment (TeamDman) [inline src/main/java/ca/teamdman/sfm/common/capabilityprovidermapper/ae2/InterfaceCapabilityProviderMapper.java]:
```suggestion
        public FluidStack drain(FluidStack resource, FluidAction action) {
```

File: src/main/java/ca/teamdman/sfm/common/capabilityprovidermapper/ae2/InterfaceCapabilityProviderMapper.java [hunk 10]
```diff
@@ -0,0 +1,446 @@
...
+
+            if (stack == null) {
+                stack = this.withBaseFluidHandler(c -> c.drain(resource, action));
+            }
+
+            return stack == null ? FluidStack.EMPTY : stack;
+        }
+
+        @Override
+        public @NotNull FluidStack drain(int maxDrain, FluidAction action) {
```
Comment (TeamDman) [inline src/main/java/ca/teamdman/sfm/common/capabilityprovidermapper/ae2/InterfaceCapabilityProviderMapper.java]:
```suggestion
        public FluidStack drain(int maxDrain, FluidAction action) {
```

File: src/main/java/ca/teamdman/sfm/common/compat/SFMCompat.java [hunk 11]
```diff
@@ -4,34 +4,54 @@
...
-                    ForgeCapabilities.ENERGY
-            );
+            ));
         }
+
+        MinecraftForge.EVENT_BUS.post(new SFMCompatGatherCapabilitiesEvent(caps));
+        CAPABILITIES.addAll(caps);
+
+        return CAPABILITIES;
     }
```
Comment (TeamDman) [inline src/main/java/ca/teamdman/sfm/common/compat/SFMCompat.java]: why are we returning a new array here?
`if (!empty) return it; else fill it and return it`
no need for set here

File: src/main/java/ca/teamdman/sfm/common/compat/SFMCompatGatherCapabilitiesEvent.java
Comment (TeamDman) [inline src/main/java/ca/teamdman/sfm/common/compat/SFMCompatGatherCapabilitiesEvent.java:1]: don't think this is necessary

any block that wants SFM compatibility should implement the IItemHandler/etc caps like normal

File: src/main/java/ca/teamdman/sfm/common/compat/SFMMekanismCompat.java [hunk 12]
```diff
@@ -14,13 +14,17 @@
...
+import net.minecraftforge.fml.ModList;
+import net.minecraftforge.fml.common.Mod;
 import net.minecraftforge.registries.DeferredRegister;
 
 import java.util.HashSet;
 import java.util.Optional;
 import java.util.Set;
 import java.util.stream.Collectors;
 
+@Mod.EventBusSubscriber
```
Comment (TeamDman) [inline src/main/java/ca/teamdman/sfm/common/compat/SFMMekanismCompat.java]: why

**Review by TeamDman — CHANGES_REQUESTED** [review]

File: src/main/java/ca/teamdman/sfm/common/compat/SFMCompat.java (hunk 11 above)
Comment (Vonr) [reply src/main/java/ca/teamdman/sfm/common/compat/SFMCompat.java]: returning the existing array is giving a mutable reference to it.

File: src/main/java/ca/teamdman/sfm/common/compat/SFMMekanismCompat.java (hunk 12 above)
Comment (Vonr) [reply src/main/java/ca/teamdman/sfm/common/compat/SFMMekanismCompat.java]: oops, artifact from converting the mekanism compat to use the event

File: src/main/java/ca/teamdman/sfm/common/compat/SFMCompatGatherCapabilitiesEvent.java
Comment (Vonr) [reply src/main/java/ca/teamdman/sfm/common/compat/SFMCompatGatherCapabilitiesEvent.java:1]: fair enough

**Review by TeamDman — APPROVED** [review]

**Pull request merged**